src/librustc_trans/mir/rvalue.rs

```diff
@@ -815,3 +816,172 @@ fn get_overflow_intrinsic(oop: OverflowOp, bcx: &Builder, ty: Ty) -> ValueRef {
+    // This yields three cases to consider:
+    // (1) if x in [f_min, f_max], steps 1, 2, and 4 do nothing and the result of fpto[su]i
+    //     is returned, which agrees with saturating conversion for inputs in that range.
+    // (2) if x > f_max, then x is larger than int_ty::MAX and step 4 correctly returns
+    //     int_ty::MAX. This holds even if f_max is rounded (i.e., if f_max < int_ty::MAX)
+    //     because in those cases, nextUp(f_max) is already larger than int_ty::MAX.
+    // (3) if x < f_min, then x is smaller than int_ty::MIN and is clamped to f_min. As shown
+    //     earlier, f_min exactly equals int_ty::MIN and therefore no fixup analogous to step 4
+    //     is needed. Instead, step 3 casts f_min to int_ty::MIN and step 4 returns this cast
+    //     result, as desired.
+    // QED.
+
+    // Step 1: Clamping. Computed as:
+    //    clamped_to_min = if f_min < x { x } else { f_min };
+    //    clamped_x = if f_max < clamped_to_min { f_max } else { clamped_to_min };
+    // Note that for x = NaN, both of the above variables become f_min.
+    let clamped_to_min = bcx.select(bcx.fcmp(llvm::RealOLT, f_min, x), x, f_min);
+    let clamped_x = bcx.select(
+                                bcx.fcmp(llvm::RealOLT, f_max, clamped_to_min),
+                                f_max,
+                                clamped_to_min
+                            );
+
+    // Step 2: NaN replacement.
+    // For unsigned types, f_min == 0.0 and therefore clamped_x is already zero.
+    // Therefore we only need to execute this step for signed integer types.
+    let clamped_x = if signed {
+        let zero = match float_ty.float_width() {
+            32 => float_bits_to_llval(ieee::Single::ZERO.to_bits()),
+            64 => float_bits_to_llval(ieee::Double::ZERO.to_bits()),
+            n => bug!("unsupported float width {}", n),
+        };
+        // LLVM has no isNaN predicate, so we use (x == x) instead
+        bcx.select(bcx.fcmp(llvm::RealOEQ, x, x), clamped_x, zero)
+    } else {
+        clamped_x
+    };
+
+    // Step 3: fpto[su]i cast
+    let cast_result = if signed {
+        bcx.fptosi(clamped_x, int_ty)
+    } else {
+        bcx.fptoui(clamped_x, int_ty)
+    };
+
+    // Step 4: f_max fixup.
+    // Note that x > f_max implies that x was clamped to f_max in step 1, and therefore the
+    // cast result is the integer equal to f_max. If the conversion from int_ty::MAX to f_max
+    // was exact, then the result of casting f_max is again int_ty::MAX, so we'd return the same
+    // value whether or not x > f_max holds. Therefore, we only need to execute this step
+    // if f_max is inexact.
+    if f_max_status.contains(Status::INEXACT) {
+        let int_max = C_big_integral(int_ty, int_max(signed, int_ty));
+        bcx.select(bcx.fcmp(llvm::RealOGT, x, f_max), int_max, cast_result)
+    } else {
+        cast_result
+    }
```

Comment:
I don't know if you're interested in micro-optimizing at this point, but if you are: it looks like this code does 4 selects for f32->i32; it would be a little shorter to take advantage of the fact that fptoui/fptosi in LLVM return undef rather than having UB, so you could do the fptosi/fptoui first, and then use selects to pick between the fptosi/fptui result value and the integer min/max/0 values in the various exceptional cases. That way, you'd always need at most 3 selects.

Reply:
I played around with this option a bit but I didn't quickly find a way to write the first two selects such that `ity::MIN` comes out if `x` is NaN. If that's not possible, i.e., a separate NaN check is always necessary, the current code is shorter (two selects) for casts such as f32 -> u16 or f64 -> f32. But perhaps I'm missing the forest for the trees? If someone can come up with a way to do three selects for signed types and two for unsigned casts, I'd gladly implement that (after I find time to address the other nits).

Reply:
Here's a way to do three selects for signed and two for unsigned, which should be generalizable to other bitwidths:

```llvm
define i32 @fptosi(float %f) {
  %t = fptosi float %f to i32
  %c0 = fcmp olt float %f, 0xC1E0000000000000 ; -0x1p31
  %c1 = fcmp oge float %f, 0x41E0000000000000 ; 0x1p31
  %c2 = fcmp uno float %f, 0.000000e+00
  %s0 = select i1 %c0, i32 -2147483648, i32 %t
  %s1 = select i1 %c1, i32 2147483647, i32 %s0
  %s2 = select i1 %c2, i32 0, i32 %s1
  ret i32 %s2
}
```

```llvm
define i32 @fptoui(float %f) {
  %t = fptoui float %f to i32
  %c0 = fcmp ule float %f, 0.000000e+00
  %c1 = fcmp oge float %f, 0x41F0000000000000 ; 0x1p32
  %s0 = select i1 %c0, i32 0, i32 %t
  %s1 = select i1 %c1, i32 -1, i32 %s0
  ret i32 %s1
}
```